docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,805 @@
+
+[float]
+==== Auto-bootstrapping in development mode
+
+If the cluster is running with a completely default configuration then it will
+automatically bootstrap based on the nodes that could be discovered within a
+short time after startup. Since nodes may not always reliably discover each
+other quickly enough this automatic bootstrapping is not always reliable and
+cannot be used in production deployments.
+
+If any of the following settings are configured then auto-bootstrapping will
+not take place, and you must configure `cluster.initial_master_nodes` as
+described in the <<modules-discovery-bootstrap-cluster,section on cluster
+bootstrapping>>:
+
+* `discovery.zen.hosts_provider`
+* `discovery.zen.ping.unicast.hosts`
+* `cluster.initial_master_nodes`
+
+[float]
+[[modules-discovery-adding-removing-nodes]]
+=== Adding and removing nodes
+
+As nodes are added or removed Elasticsearch maintains an optimal level of fault
+tolerance by updating the cluster's _voting configuration_, which is the set of
+master-eligible nodes whose responses are counted when making decisions such as
+electing a new master or committing a new cluster state.
+
+It is recommended to have a small and fixed number of master-eligible nodes in a
+cluster, and to scale the cluster up and down by adding and removing
+master-ineligible nodes only. However there are situations in which it may be
+desirable to add or remove some master-eligible nodes to or from a cluster.
+
+If you wish to add some master-eligible nodes to your cluster, simply configure
+the new nodes to find the existing cluster and start them up. Elasticsearch will
+add the new nodes to the voting configuration if it is appropriate to do so.
+
+When removing master-eligible nodes, it is important not to remove too many all
+at the same time. For instance, if there are currently seven master-eligible
+nodes and you wish to reduce this to three, it is not possible simply to stop
+four of the nodes at once: to do so would leave only three nodes remaining,
+which is less than half of the voting configuration, which means the cluster
+cannot take any further actions.
+
+As long as there are at least three master-eligible nodes in the cluster, as a
+general rule it is best to remove nodes one-at-a-time, allowing enough time for
+the cluster to <<modules-discovery-quorums,auto-adjust>> the voting
+configuration and adapt the fault tolerance level to the new set of nodes.
+
+If there are only two master-eligible nodes remaining then neither node can be
+safely removed since both are required to reliably make progress, so you must
+first inform Elasticsearch that one of the nodes should not be part of the voting
+configuration, and that the voting power should instead be given to other nodes,
+allowing the excluded node to be taken offline without preventing the other node
+from making progress. A node which is added to a voting configuration exclusion
```

Comment:
The sentence starting with `A node` is confusing. Maybe unpack a bit to clarify?

Reply:
Tried to improve this in 2fdb92f7eb997d9255b1bf821395c5b6a1f5acc2.